drivers: ssd16xx: allow to use WS from OTP memory, add support for GDEH0154D67 display

jfischer-no opened the pull request:
Add support for GDEH0154D67 display
Add support to load default WS from OTP
Add temperature sensor support

vanwinkeljan commented on drivers/display/ssd16xx.c:
```diff
@@ -62,42 +62,38 @@ struct ssd16xx_data {
-	if (err < 0) {
-		return err;
+	if (spi_write(driver->spi_dev, &driver->spi_config, &buf_set)) {
+		return -EIO;
```
What is the reason to change this and not just return the actual error code to the caller?

jfischer-phytec-iot replied:
Do not forward minor spi_write return value and
**save few lines of code** (`@retval 0 If successful, negative errno code otherwise.`)
Treat undefined positive return value as an error.

vanwinkeljan replied:
> edit: more precise, unspecified return value

You have seen them, if this the case sound like a bug in a SPI driver?

jfischer-phytec-iot replied:
oh no, it is just unspecified, but we have it on many places. The change above is more optical.

jfischer-phytec-iot commented:
needs some rework, postpone to 2.4.0

jfischer-no committed:
shields: add support for GDEH0154D67 display

Add support for GDEH0154D67 display.

Signed-off-by: Johann Fischer <j.fischer@phytec.de>

jfischer-no committed:
drivers: ssd16xx: support to load default WS from OTP

Add support to load default WS from OTP.

Signed-off-by: Johann Fischer <j.fischer@phytec.de>

jfischer-no committed:
drivers: ssd16xx: add temperature sensor support

Appropriate WS can be loaded automatically if
the display controller has integrated temperature
sensor or an external sensor is connected.

Signed-off-by: Johann Fischer <j.fischer@phytec.de>

vanwinkeljan commented on drivers/display/ssd16xx.c:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2018 PHYTEC Messtechnik GmbH
+ * Copyright (c) 2018-2018 PHYTEC Messtechnik GmbH
```
Typo, 2020?

jfischer-phytec-iot replied:
thanks, fixed

weinholtendian commented:
I'm working with some displays that have values pre-programmed in OTP, so I'm interested in these changes and will be able to try them out with some different displays soon.

@jfischer-phytec-iot  Do you already have plans to update the display with temperature readings from a sensor that the application has access to? Often the MCU has a temperature sensor that could be used.

Right now I'm having some trouble getting the driver to work at all on an nRF52840; the display just doesn't update. Tried with a gdeh0213b1 and a gdeh0213b72.

jfischer-phytec-iot commented:
> @jfischer-phytec-iot Do you already have plans to update the display with temperature readings from a sensor that the application has access to? Often the MCU has a temperature sensor that could be used.

Generally it is not a good idea :-), there is no API for that and enhance the API for this very specific case seems not necessary for me. The driver self could use sensor API to obtain the values and pass it to the controller, I thought about it. But, my experience with the GD displays is that those without internal temperature sensor usually do not have valid set of WS in the OTP. If you are developing a new product use a display with internal temperature sensor or connect one to the display controller's I2C interface.  The display you mentioned (gdeh0213b1 and a gdeh0213b72) are long time EOL, and the way these display are treated in the samples here is not how one would do it in a product.

weinholtendian commented:
> If you are developing a new product use a display with internal temperature sensor or connect one to the display controller's I2C interface.

Thank you very much for this advice! Then it does indeed not make sense to add an API for this.

>  The display you mentioned (gdeh0213b1 and a gdeh0213b72) are long time EOL, 

I'm using displays from GoodDisplay just to get a baseline with known working components. The final product will be using displays from another manufacturer that also uses an ssd16xx controller. They use OTP but have some minor electrical interface differences, so I will be testing this change, but don't know when testing can happen.

> and the way these display are treated in the samples here is not how one would do it in a product.

Are you referring to the fact that the display is not powered off or put to sleep? (I don't have a good reference now, but one document that I ran across mentioned possible damage if the display is kept in a powered on state for too long).

jfischer-phytec-iot commented:
> > and the way these display are treated in the samples here is not how one would do it in a product.
> 
> Are you referring to the fact that the display is not powered off or put to sleep? (I don't have a good reference now, but one document that I ran across mentioned possible damage if the display is kept in a powered on state for too long).

One would not update the display like the samples do it. Faster WS (from device tree) for gdeh0213b1 are IMHO not right, typically you would update the display with a WS from OTP and put the SoC in a sleep mode. Controller's Deep Sleep mode is not supported because the Sleep mode current was acceptable and there is probably a issue with the controller's circuit or controller self. We find out that display controller emits small amount of energy over BUSY pin at the end of the update sequence. This leads to a voltage overshoot on the power line of a low power design, that is why reel board has LED on the BUSY pin of the display controller.
